summingbird-storm/src/main/scala/com/twitter/summingbird/storm/collector/AggregatorOutputCollector.scala

```diff
@@ -75,43 +80,24 @@ class AggregatorOutputCollector[K, V: Semigroup](
    This is a wrapper method to call the emit with appropriate signature
    based on the arguments.
   */
-  private def callEmit(messageIds: Option[TraversableOnce[AnyRef]], list: JList[AnyRef], stream: String): JList[Integer] = {
-    (messageIds, stream.isEmpty) match {
-      case (None, true) => in.emit(list)
-      case (None, false) => in.emit(stream, list)
-      case (Some(ids), true) => in.emit(list, ids)
-      case (Some(ids), false) => in.emit(stream, list, ids)
+  private def callEmit(tuple: JList[AnyRef], messageIds: TraversableOnce[AnyRef], stream: String): JList[Integer] = {
+    (messageIds.isEmpty, stream.isEmpty) match {
+      case (true, true) => in.emit(tuple)
+      case (true, false) => in.emit(stream, tuple)
+      case (false, true) => in.emit(tuple, messageIds)
+      case (false, false) => in.emit(stream, tuple, messageIds)
     }
   }
 
   /*
   Method wraps the adding the tuple to the spoutCache along with adding the corresponding
   messageId to the messageId Tracker.
    */
-  private def add(tuple: (K, V), streamid: String, messageId: Option[AnyRef] = None): Future[Map[K, V]] = {
-    if (messageId.isDefined)
-      trackMessageId(tuple, messageId.get, streamid)
-    addToCache(tuple, streamid)
-  }
-
-  /*
-   * As there are separate summers for each stream. This method takes care of the stream lookup
-   * and adding the tuple to the corresponding cache.
-   */
-  private def addToCache(tuple: (K, V), streamid: String): Future[Map[K, V]] = {
-    cacheByStreamId.getOrElseUpdate(streamid, summerBuilder.getSummer[K, V](implicitly[Semigroup[V]]))
-      .add(tuple)
-  }
-
-  /**
-   * The lookup of stream is followed by the lookup on summerShard happens.
-   * The messageId is added to the corresponding group.
-   * All the messageIds are sent along with the crushed tuple.
-   */
-  private def trackMessageId(tuple: (K, V), o: AnyRef, s: String): Unit = {
-    val messageIdTracker = streamMessageIdTracker.getOrElseUpdate(s, MMap[Int, MList[Object]]())
-    val messageIds = messageIdTracker.getOrElseUpdate(summerShards.summerIdFor(tuple._1), MList())
-    messageIds += o
+  private def add(tuple: (K, V), streamId: String, messageId: Option[AnyRef] = None): Future[Map[K, (Seq[Object], V)]] = {
+    val buffer = messageId.map { id => ListBuffer(id) }.getOrElse(Nil)
+    val (k, v) = tuple
+    cacheByStreamId.getOrElseUpdate(streamId, summerBuilder.getSummer[K, (Seq[Object], V)](implicitly[Semigroup[(Seq[Object], V)]]))
```

Comment:
Seq semigroup may not be that great. Without using `sumOption` it is `O(N^2)` to aggregate.

Have you looked at `Batched`: https://github.com/twitter/algebird/blob/develop/algebird-core/src/main/scala/com/twitter/algebird/Batched.scala that deals with this?

We may want to wrap `Seq[AnyRef]` with `Batched[Seq[AnyRef]]`.

Also, can we have a private method like:
```scala
private def getSummer =
  summerBuilder.getSummer[K, (Seq[Object], V)](implicitly[Semigroup[(Seq[Object], V)]])
```
and call that here rather than the long line?